### PR TITLE
reduce credit update messages sent to guest

### LINKED ISSUE #317: vsock: reduce credit update messages sent to the guest
The virtio-vsock driver and vhost-vsock device emulation in the Linux kernel have been optimized [1] to reduce the number of credit update messages exchanged [2].
The device in vhost-user-vsock behaves like the device in Linux before those changes, sending a credit update message after each read, even of a single byte.
We should reduce the credit update messages sent to the guest doing something similar of what Linux driver and device do.
​
[1] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=c69e6eafff5f725bc29dcb8b52b6782dca8ea8a2
[2] https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=b89d882dc9fc279c8acbf1df71d51b22394186d5

### PR DESCRIPTION
**Pull request by ramyak-mehra**
earlier we were sending credit update messages to guest everytime data was read, now we only send it only if available space is less than which is TX_BUFF_SIZE / 4.
Signed-off-by: Ramyak Mehra <rmehra_be19@thapar.edu>

### Summary of the PR

Fixes #317 
### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: crates/vsock/src/vhu_vsock.rs
Comment (stefano-garzarella) [inline crates/vsock/src/vhu_vsock.rs:42]: Can you explain why use `CONN_TX_BUF_SIZE / 4`?

Maybe we can use the same value we used in Linux, but if you prefer to use that value, please rebase on top of #324

**Review by stefano-garzarella — COMMENTED** [review]
Can you also provides some performance numbers (e.g. using `iperf-vsock`)?

File: crates/vsock/src/vhu_vsock.rs
Comment (ramyak-mehra) [reply crates/vsock/src/vhu_vsock.rs:42]: I was gonna add the link to it, in the Linux codebase the value is 1/4th of thr buffer size that's why I chose it

Comment (stefano-garzarella) [reply crates/vsock/src/vhu_vsock.rs]: Sorry, I can't see it in the Linux codebase, can you put a link?

Comment (stefano-garzarella) [reply crates/vsock/src/vhu_vsock.rs]: I think it is just a case that it is 1/4 of the default value of buf_alloc, but if buf_alloc is changed the threashold remains fixed. I think it makes sense to use the same value, or do some experiments to show that choosing 1/4 is okay.

**Comment by stefano-garzarella** [comment]
I'd suggest merging the commit that fixed the rebase with the original one (at least before merging this PR ;-)

File: crates/vsock/src/vhu_vsock.rs
Comment (ramyak-mehra) [reply crates/vsock/src/vhu_vsock.rs]: Yes the value is 1/4 of the max buf_alloc. The reason i kept it variable and actually dependent upon buf_alloc is because since in our case the allocated buffer could be of any size, assume it's very small and we have the threshold as quite high we would be again sending the updates on every read sort off. If we had a range for that then we could still try to make it a fixed value. I tried with different values 1/4 seems fine and produced good results. I have the numbers and was gonna post that as part of the benchmarks

**Comment by ramyak-mehra** [comment]
> I'd suggest merging the commit that fixed the rebase with the original one (at least before merging this PR ;-)

Yeah, I was gonna add the benchmarking results and then just do a squash and merge.

File: crates/vsock/src/vsock_conn.rs
Comment (stefano-garzarella) [inline crates/vsock/src/vsock_conn.rs:323]: Since we are requesting `self.tx_buf.get_buf_size()` here, perhaps we can avoid introducing `credit_update_threashold` and calculating it each time.

File: crates/vsock/src/vhu_vsock.rs
Comment (stefano-garzarella) [reply crates/vsock/src/vhu_vsock.rs]: Okay, it makes sense to me ;-)

File: crates/vsock/src/vsock_conn.rs
Comment (JonathanWoollett-Light) [inline crates/vsock/src/vsock_conn.rs]: Personal opinion, I find `wrapping_sub` to be more clear, than [`Wrapping`](https://doc.rust-lang.org/std/num/struct.Wrapping.html).
```suggestion
            let free_space = CONN_TX_BUF_SIZE.wrapping_sub(self.fwd_cnt - self.last_fwd_cnt);
```

File: crates/vsock/src/vsock_conn.rs
Comment (JonathanWoollett-Light) [inline crates/vsock/src/vsock_conn.rs]: Personal opinion, I find `wrapping_sub` to be more readable and simple, than [`Wrapping`](https://doc.rust-lang.org/std/num/struct.Wrapping.html).
```suggestion
            let free_space = CONN_TX_BUF_SIZE.wrapping_sub(self.fwd_cnt - self.last_fwd_cnt);
            if free_space < CREDIT_UPDATE_THRESHOLD {
                self.rx_queue.enqueue(RxOps::CreditUpdate);
            }
```

File: crates/vsock/src/vsock_conn.rs
Comment (stefano-garzarella) [inline crates/vsock/src/vsock_conn.rs:319]: What about storing `self.tx_buf.get_buf_size()` in a local variable?

**Review by stefano-garzarella — DISMISSED** [review]
LGTM, I just left a minor comment and maybe we can update the commit description replacing `CONN_TX_BUF_SIZE/4` with `TX buffer size`.

**Comment by stefano-garzarella** [comment]
And we should mention that this fixes #317, here in the PR description or even better in the commit description, adding `Fixes: #317`

**Comment by stefano-garzarella** [comment]
@ramyak-mehra now the CI is working, can you fix the commit message to make it happy?

This is the error reported:
```
E    AssertionError: For commit '8ccace1', message line 'Note: I did 3 runs of 10 secs and took average of all, they are relative results.' exceeds 75 chars. Please keep it shorter or split it in multiple lines.
E    assert 81 <= 75
E    +  where 81 = len('Note: I did 3 runs of 10 secs and took average of all, they are relative results.')
```

**Comment by stefano-garzarella** [comment]
> @ramyak-mehra now the CI is working, can you fix the commit message to make it happy?
> 
> This is the error reported:
> 
> ```
> E    AssertionError: For commit '8ccace1', message line 'Note: I did 3 runs of 10 secs and took average of all, they are relative results.' exceeds 75 chars. Please keep it shorter or split it in multiple lines.
> E    assert 81 <= 75
> E    +  where 81 = len('Note: I did 3 runs of 10 secs and took average of all, they are relative results.')
> ```

Since you're there, can you also add `vsock:` prefix to the title?

**Review by stefano-garzarella — APPROVED** [review]
For other reviewers: this PR depends on #324, so the first commit is from that PR and we should merge this PR after it.

**Comment by vireshk** [comment]
> earlier we were sending credit update messages to guest everytime data was read, now we only send it only if 

The commits log isn't properly formatted. It should be a proper english sentence. Like first letter should be "Earlier" instead of earlier. Subject should have "vsock: " and a blank line before Signed-off line.

Please rebase this branch, now that other commit is already merged, which will simplify the review here.

**Comment by vireshk** [comment]
This needs to be rebased now

**Comment by vireshk** [comment]
Any update on this ? Can we get this fixed and merged soon ?

**Comment by ramyak-mehra** [comment]
Sorry for the delay I was caught up with some college work. Will update and push it in couple of hours

**Comment by stefano-garzarella** [comment]
@vireshk @ramyak-mehra I just rebased this commit on current `main` and fixed something in the commit description to make DCO happy and address Viresh's comments.

Since we are going to do a release soon, we should merge this ASAP.

**Review by stefano-garzarella — APPROVED** [review]

**Review by vireshk — APPROVED** [review]

**Pull request merged**
